[Feat] 2차 세미나 과제 (기본, 도전)

JinUng41 opened the pull request:
## What is the PR? 🔍
<!-- 아래 리스트를 지우고, 작업하게 된 배경을 적어주세요. -->
 - 2차 세미나 과제 (기본, 도전)
 - 로그인, 닉네임, 환영 화면 구현하기

## Changes 📝
<!-- 작업 내용 및 덧붙이고 싶은 내용이 있다면! -->
### 입력에 대한 에러 처리
- 아이디(이메일), 비밀번호, 닉네임을 정규식을 통해 검증하는 부분을 에러를 `throw`할 수 있도록 구현
- 에러 케이스를 정의하고 Alert를 띄울 수 있도록 하였습니다.
<details>
<summary>AppError 코드</summary>

```swift
enum AppError: Error, CustomStringConvertible {
    case login(error: LoginError)
    case nickname
    
    var description: String {
        switch self {
        case .login(let error):
            switch error {
            case .invalidID:
                "아이디 에러"
            case .invalidPW:
                "비밀번호 에러"
            }
        case .nickname:
            "닉네임 에러"
        }
    }
    
    var message: String {
        switch self {
        case .login(let error):
            error.description
        case .nickname:
            """
            올바른 닉네임 형식이 아닙니다.
            1~10자 이내 한글만 사용해 주세요.
            """
        }
    }
}

// MARK: - Error Case

extension AppError {
    enum LoginError: CustomStringConvertible {
        case invalidID, invalidPW
        
        var description: String {
            switch self {
            case .invalidID:
                "올바른 이메일 형식이 아닙니다."
            case .invalidPW:
            """
            올바른 비밀번호 형식이 아닙니다.
            8~10자 이내 영어, 특수기호만 사용해 주세요.
            """
            }
        }
    }
}

```
</details>

<details>
<summary>AlertShowable 코드</summary>

```swift
protocol AlertShowable {
    func showAlert(title: String, message: String)
}

extension AlertShowable where Self: UIViewController {
    func showAlert(title: String, message: String) {
        let alertController = UIAlertController(title: title, message: message, preferredStyle: .alert)
        let confirmAction = UIAlertAction(title: "확인", style: .default)
        alertController.addAction(confirmAction)
        present(alertController, animated: true)
    }
}

```
</details>

## Screenshot 📷
<!-- 작업한 화면이 있다면 스크린 샷으로 첨부해주세요. -->

|    오류처리    |  정상동작   |
| :-------------: | :----------: |
| <img src = "https://github.com/NOW-SOPT-iOS-Part/KimJinUng-assignment/assets/91656206/f13c932d-0367-4ab6-8176-371ebd35d4bf" width ="250">| <img src = "https://github.com/NOW-SOPT-iOS-Part/KimJinUng-assignment/assets/91656206/5eff4af2-76fe-4678-8f70-6a1e5d7d98c2" width ="250">|

## To Reviewers 🙏
<!-- 리뷰어에게 주목했으면 하는 점 or 바라는 점을 적어주세요. -->
- 커스텀 델리게이트 패턴을 이용한 데이터 전달 (닉네임)
- 하드 코딩을 줄이기 위한 Constants 구현
- MARK 주석 및 Then 적극 사용
- 익스텐션을 이용한 코드 간략화
 
## Related Issues 💭
<!-- 작업한 이슈번호를 # 뒤에 붙여주세요. 수고했습니다~* -->
- Resolved: #1

JinUng41 committed:
#1 - [Config] Then, Snapkit 추가

JinUng41 committed:
#1 - [Config] 로고 이미지 및 컬러 에셋 추가

JinUng41 committed:
#1 - [Config] 버튼 이미지 추가

JinUng41 committed:
#1 - [Config] Pretendard 폰트 추가

JinUng41 committed:
#1 - [Feat] extension 추가

- UIFont
- UITextField
- UIView

JinUng41 committed:
#1 - [Feat] 로그인 화면 1차 UI 구현

- 기능 구현 아직
- 텍스트 필드에 기능 버튼 추가로 구현 아직

JinUng41 committed:
#1 - [Feat] 로그인 화면 텍스트필드의 버튼 기능 구현

- 텍스트 필드의 버튼 기능 (필드 내용 삭제 및 비밀번호 표시 구현)

JinUng41 committed:
#1 - [Feat] UIColor Extension 추가

- GrayScale의 Hex에 보다 쉽게 대응하기 위한 방법 구현

JinUng41 committed:
#1 - [Feat] 3가지 모델 정의

- 아이디, 비밀번호, 닉네임 검증 과정에서 처리할 오류
- 검증식 종류
- 상수

JinUng41 committed:
#1 - [Refactor] 익스텐션 수정 및 빈줄 삭제

JinUng41 committed:
#1 - [Feat] 정규식 검증 및 오류 경고창 띄우는 기능 추가

JinUng41 committed:
#1 - [Feat] 폴더 구조 변경 및 로그인 기능 구현

JinUng41 committed:
#1 - [Chore] 불필요한 구문 삭제

JinUng41 committed:
#1 - [Refactor] 버튼의 레이어 설정 방식 수정

JinUng41 committed:
#1 - [Feat] 닉네임 입력 화면 UI 및 기능 구현

JinUng41 committed:
#1 - [Feat] 로그인 시 아이디 또는 닉네임을 보여주는 화면 UI 및 기능 구현

JinUng41 committed:
#1 - [Refactor] 웰컴 화면 네비게이션 뒤로가기 버튼 숨기기

meltsplit commented on 34th-Sopt-Assignment/34th-Sopt-Assignment/Common/Error/AppError.swift:
```diff
@@ -0,0 +1,59 @@
+
+import Foundation
+
+enum AppError: Error, CustomStringConvertible {
+    case login(error: LoginError)
+    case nickname
+    
+    var description: String {
+        switch self {
+        case .login(let error):
+            switch error {
+            case .invalidID:
+                "아이디 에러"
+            case .invalidPW:
+                "비밀번호 에러"
+            }
+        case .nickname:
+            "닉네임 에러"
+        }
+    }
+    
+    var message: String {
+        switch self {
+        case .login(let error):
+            error.description
+        case .nickname:
+            """
+            올바른 닉네임 형식이 아닙니다.
+            1~10자 이내 한글만 사용해 주세요.
+            """
+        }
+    }
+}
```
CustomStringConvertible 프로토콜 사용하는 거 좋은 것 같습니다! 
print문 찍을 때 보기 쉽게 하기 위해서 인가염?

JinUng41 replied:
`CustomStringConvertible`을 채택하여 에러를 catch 했을 때, _AlertShowable_ 의 showAlert 메서드에서 보다 사용하기 쉽게 하기 위함이었습니다.

meltsplit replied:
아항 네네 description과 message 프로퍼티가 유사하지만 다른듯하여 여쭤봤습니닷!

meltsplit commented on 34th-Sopt-Assignment/34th-Sopt-Assignment/Common/Error/AppError.swift:
```diff
@@ -0,0 +1,59 @@
+
+import Foundation
+
+enum AppError: Error, CustomStringConvertible {
+    case login(error: LoginError)
+    case nickname
+    
+    var description: String {
+        switch self {
+        case .login(let error):
+            switch error {
+            case .invalidID:
+                "아이디 에러"
+            case .invalidPW:
+                "비밀번호 에러"
+            }
+        case .nickname:
+            "닉네임 에러"
+        }
+    }
+    
+    var message: String {
+        switch self {
+        case .login(let error):
+            error.description
+        case .nickname:
+            """
+            올바른 닉네임 형식이 아닙니다.
+            1~10자 이내 한글만 사용해 주세요.
+            """
+        }
+    }
+}
```
AppError 내부에 LoginError 두신게 인상깊네요!
진웅님은 앱 내부의 모든 에러를 AppError에 계층적으로 선언하는걸 선호하시는 편이신가용?
Error를 계층적으로 쌓으신 이유가 궁금해요 :)

JinUng41 replied:
사실 이렇게 계층적으로 오류에 대한 구현은 처음입니다. ㅎㅎ
원래 저의 스타일이었다면, AppError 내에 invalidID, invalidPW, nickname 전부 있었을 것 같은데요.
이번에 구현을 해보면서 좀 더 생각해 봤던 점은 좀 더 기능에 집중하여 세분화된 에러 모델을 만드는 것이었고, 현재의 코드 처럼 구현이 되었습니다.
'앱의 전체 에러 > 앱의 기능 에러 (로그인) > 아이디 오류 또는 비밀번호 오류'의 흐름 순으로 생각을 가지고 구현했을 때, 보다 이해하기 쉬울 것 같다는 생각이 들었습니다.
이에 멜트스플릿의 생각은 어떠신지 들어보고 싶습니다.

meltsplit replied:
저도 최근에 에러 모델을 어떻게 구축해야할지 고민이 많아서 여쭤봤습니다.

계층적 에러의 한가지 우려되는 점은 아직 화면이 하나밖에 없어서 쉽게 이해가 되지만, 
화면이 여러개가 되고 에러 수가 많아질때 에러를 어떤 계층으로 설계해야할 지 감이 잘 오진 않는 것 입니당

저는  **에러를 어디서 잡을지**에 따라 에러모델의 형태도 달라진다고 생각해욥.
제가 하고 있는 플젝에서는 에러를 아키텍처(MVC or MVVM or 클린아키텍처.. etc)에서 계층마다 잡는 것이 바람직하다고 생각하여 에러를 계층적으로 쌓지 않고, 아키텍처의 계층마다 에러 모델을 구축하여 설계하고 있습니다!

meltsplit commented on 34th-Sopt-Assignment/34th-Sopt-Assignment/Common/Protocols/AlertShowable.swift:
```diff
@@ -0,0 +1,21 @@
+
+import UIKit
+
+protocol AlertShowable {
+    func showAlert(title: String, message: String)
+}
+
+extension AlertShowable where Self: UIViewController {
+    func showAlert(title: String, message: String) {
+        let alertController = UIAlertController(title: title, message: message, preferredStyle: .alert)
+        let confirmAction = UIAlertAction(title: "확인", style: .default)
+        alertController.addAction(confirmAction)
+        present(alertController, animated: true)
+    }
+}
```
Protocol, Extension, Where문을 잘 활용하시는 것 같아요!
배워갑니당!

JinUng41 replied:
감사합니다! 🙏

codeJiwon replied:
혹시 여기서 where문을 사용한 이유 여쭤봐도 될까요?? UIViewController를 상속하는 경우에만 extension 적용시키려는 의도셨을까요?

meltsplit commented on 34th-Sopt-Assignment/34th-Sopt-Assignment/Controllers/Login/MakeNicknameViewController.swift:
```diff
@@ -0,0 +1,159 @@
+            $0.layer.borderWidth = 0
+            $0.isEnabled = true
+        }
+    }
```
저는 이런 같은 상황일 때 코드를 아래와 같이 구현해요!

```swift
private func nicknameTextFieldEditingChanged(_ sender: UITextField) {
        let isEnabled = sender.text?.isEmpty == false
        updateSaveButtonUI(isEnabled)
    }

private func updateSaveButtonUI(_ isEnabled: Bool) {
    let titleColor = isEnabled ? .basicWhite : .gray4
    let backgroundColor = isEnabled ? .brandRed : .basicBlack

    //...생략

    saveButton.setTitleColor(titleColor, for: .normal)
    saveButton.backgroundColor = backgroundColor
}
```
저와 진웅님이 다른 점은 아래와 같을 것 같네용
1. guard 문 사용 여부
2. UI 관련 함수 분리 여부
3. 삼항연산자 사용여부

어떤게 더 좋은 방법인지 저두 잘 모르겠네여 o0o..
진웅님의 의견을 듣고시퍼요!

seungchan2 replied:
석우님 좋네요

JinUng41 replied:
저는 UI 작업 코드가 멜트스플릿의 코드와 제 코드 모두 분리가 되어있다고 생각합니다.
하지만 역시 비슷한 내용이 반복된다는 점에서 제 코드 보다는 삼항연산자를 활용하는 방법이 더 나아보인다고 판단되네요.
따라서 해당 부분과 관련 부분들을 삼항 연산자로 수정하였습니다.
의견 감사합니다!

meltsplit commented on 34th-Sopt-Assignment/34th-Sopt-Assignment/Controllers/Login/MakeNicknameViewController.swift:
```diff
@@ -0,0 +1,159 @@
+            let error = error as! AppError
+            showAlert(title: "\(error)", message: "\(error.message)")
+        }
+    }
```
`do catch`문 사용 너무 잘 하신 것 같아요!
하지만 강제언래핑이 다소 위험해보입니다,,
강제언래핑을 사용한 이유 있으실까욥?

JinUng41 replied:
제가 직접 정의한 checkNickname 메서드에 의해 에러가 발생할 수 있기 때문에 강제 언래핑을 사용하였는데요.
하지만 역시 강제 언래핑은 런타임 에러를 발생시킬 수 있기 때문에, 사용하지 않는 것이 좋아보입니다. 
따라서 아래와 같이 catch let 구문을 활용하여 수정해 보았습니다.

```swift
} catch let appError as AppError {
    showAlert(title: "\(appError)", message: "\(appError.message)")
} catch {
    print("\(error.localizedDescription)")
}
```

meltsplit commented on 34th-Sopt-Assignment/34th-Sopt-Assignment/Controllers/Login/MakeNicknameViewController.swift:
```diff
@@ -0,0 +1,159 @@
+        }
+        return input
+    }
+}
```
검증하는 함수로 `throws`문 사용 너무 좋은 것 같아요!
순수함수를 위해 아래와 같이 사용해 보는 건 어떨까요?

```swift
private func checkNickname(input: String?) throws -> String {
        guard let input = input,
              checkFrom(input: input, regex: .nickname)
        else {
            throw AppError.nickname
        }
        return input
    }
```

JinUng41 replied:
네 말씀 감사합니다. 코드에 반영하였습니다!

meltsplit commented on 34th-Sopt-Assignment/34th-Sopt-Assignment/Controllers/Login/MakeNicknameViewController.swift:
```diff
@@ -0,0 +1,159 @@
+        saveButton.snp.makeConstraints {
+            $0.bottom.equalTo(safeArea.snp.bottom).offset(-10)
+            $0.leading.trailing.height.equalTo(nicknameTextField)
+        }
```
leading과 trailing을 합친 `호리엔탈 엣지`도 있습니다!
```swift
saveButton.snp.makeConstraints {
            $0.bottom.equalTo(safeArea.snp.bottom).offset(-10)
            $0.호리젠탈엣지.equalTo(nicknameTextField)
            $0.height.equalTo(nicknameTextField)
        }
```

JinUng41 replied:
오, 이 부분은 미처 몰랐던 부분입니다. 알려주셔서 감사해요! 🙏

meltsplit commented on 34th-Sopt-Assignment/34th-Sopt-Assignment/Common/Constants/Constants.swift:
```diff
@@ -0,0 +1,40 @@
+
+// MARK: - UI
+
+extension Constants {
+    struct UI {
+        static let cornerRadius: CGFloat = 3
+        static let textFieldAndButtonHeight = 52
+    }
+}
```
저는 UI 관련 상수는 따로 constant에 묶지 않는편이에요!
디자인시스템이 명확하게 구축되어있지 않으면 컴포넌트마다 매번 바뀔수 있는 값들이라 
각 뷰에 쓰는 것을 선호한답니당~

JinUng41 replied:
Figma로 UI를 확인해 가면서, UI를 구현했을 때 해당 두 프로퍼티는 특히 텍스트필드와 버튼에 있어 자명하게 쓰인다고 판단하고 구현해 보았습니다.
말씀 감사합니다~

meltsplit commented on 34th-Sopt-Assignment/34th-Sopt-Assignment/Common/Constants/Constants.swift:
```diff
@@ -0,0 +1,40 @@
+
+// MARK: - Image
+
+extension Constants {
+    struct Image {
+        static let eye_slash = "eye_slash"
+        static let eye = "eye"
+        static let logo = "logo"
+        static let x_circle = "x_circle"
+    }
+}
```
Xcode 15부턴 [<UIImage(_ resource: ImageResource)>](https://developer.apple.com/documentation/uikit/uiimage/4176594-init)이 사용가능해서
저는 Image name을 Constants에 따로 묶지 않고 
ImageResource가 자동적으로 생성해주는 asseet의 번들네임? 을 사용합니다!
엄청 편해요!

JinUng41 replied:
몰랐던 부분이네요. 앞으로 외부 이미지를 보다 편하게 사용 가능할 것 같습니다. 감사합니다! 🙏
따라서 해당 내용은 삭제하였습니다.

meltsplit reviewed (approved):
안녕하세요! 🤖 코드리뷰봇 멜트스플릿입니다ㅎㅎ

과제 보니 코드의 품질이 매우 좋으신것같아요~!
특히 코드컨벤션을 잘 지켜주시는 것과 swift 문법 이해도가 높으신 게 큰 장점이신 것 같아요!!
저도 많이 배워갑니다.

코드를 보며 궁금했던 점 몇가지 남기고 갑니다
수고하셨어용 :)

seungchan2 commented on 34th-Sopt-Assignment/34th-Sopt-Assignment/Common/Extensions/UIButton+Extension.swift:
```diff
@@ -0,0 +1,32 @@
+import UIKit
+
+extension UIButton {
+    func setTitle(title: String, titleColor: UIColor = .white, font: UIFont? = nil) {
```
AppleGuideLine 참고하면 세개 이상의 파라미터를 받을 때 줄바꿈 하도록 권장하더라구요!
```
func setTitle(title: String,
 titleColor: UIColor = .white,
 font: UIFont? = nil)
```

JinUng41 replied:
아하! 그렇군요.
사실 저 같은 경우에는 130자를 제한으로 두고, 130자를 넘기게 되면 그 때 개행 처리를 하는 코드 컨벤션을 가져가려 하는데 리뷰어의 생각은 어떠신지 궁금합니다!

seungchan2 replied:
130자까지 세어보지는 않아서용 ... 죄송합니다...
일단 파라미터 세개 이상 받을 때, 저는 줄바꿈합니다 ..

seungchan2 commented on 34th-Sopt-Assignment/34th-Sopt-Assignment/Controllers/Login/MakeNicknameViewController.swift:
```diff
@@ -0,0 +1,159 @@
+    func configure(nickname: String)
+}
+
+final class MakeNicknameViewController: UIViewController, RegexCheckable, AlertShowable {
```
해당 프로토콜도 따로 분리하는게 좋아보이는데 어떻게 생각하세요?

JinUng41 replied:
_RegexCheckable_ 과 _AlertShowable_ 의 경우, 현재 구현에서는 프로토콜의 실제 내용을 정의 하지 않고 기본 구현만을 사용하도록 하고 있습니다. 해당 경우에도 extension으로 분리하는게 좋을지 의문입니다.
이에 리뷰어의 생각을 여쭤보고 싶어요!

seungchan2 replied:
앗 그러면 상관없어보이네요 !
typealias로 묶어도 가독성 좋을 것 같긴하네요!

seungchan2 commented on 34th-Sopt-Assignment/34th-Sopt-Assignment/Controllers/Login/MakeNicknameViewController.swift:
```diff
@@ -0,0 +1,159 @@
+    }
+}
+
+extension MakeNicknameViewController {
```
저는 아래 스타일을 선호합니다. 그렇다구요.
```
private extension { 
func setUI() {
   }
}
```

JinUng41 replied:
오, 현재 저의 코드에서 UI를 설정하는 부분이기 때문에, 독립적인 느낌으로 private extension을 사용하는 것이 더 좋아보입니다. 
알려주셔서 감사합니다! 🙏

seungchan2 reviewed (commented):
야미하네요

seungchan2 commented on 34th-Sopt-Assignment/34th-Sopt-Assignment/Common/Error/AppError.swift:
```diff
@@ -0,0 +1,59 @@
+    case login(error: LoginError)
+    case nickname
+    
+    var description: String {
+        switch self {
+        case .login(let error):
+            switch error {
+            case .invalidID:
+                "아이디 에러"
+            case .invalidPW:
+                "비밀번호 에러"
+            }
+        case .nickname:
+            "닉네임 에러"
+        }
+    }
```
제가 좋아하는 코드스타일입니다 👍🏻

JinUng41 replied:
감사합니다!

seungchan2 commented on 34th-Sopt-Assignment/34th-Sopt-Assignment/Common/Extensions/UIFont+Extension.swift:
```diff
@@ -0,0 +1,32 @@
+        
+        var font: String {
+            switch self {
+            case .one: "Pretendard-Thin"
+            case .two: "Pretendard-ExtraLight"
+            case .three: "Pretendard-Light"
+            case .four: "Pretendard-Regular"
+            case .five: "Pretendard-Medium"
+            case .six: "Pretendard-SemiBold"
+            case .seven: "Pretendard-Bold"
+            case .eight: "Pretendard-ExtraBold"
+            case .nine: "Pretendard-Black"
```
네이밍을 thin, bold 이렇게 짓는건 어떠신가요?

JinUng41 replied:
Figma를 확인해 보았을 때, 폰트에 대하여 가중치를 숫자 100단위로 보여주기 때문에 one~nine까지로 구현해보았습니다.

seungchan2 replied:
저는 종종 폰트명이 헷갈려서 ,, thin, light 이렇게 선언하는걸 선호합니다!
추후 협업간에 논의도 중요하겠네요!

seungchan2 commented on 34th-Sopt-Assignment/34th-Sopt-Assignment/Controllers/Login/LoginViewController.swift:
```diff
@@ -0,0 +1,406 @@
+        enableLoginButton()
+    }
+    
+    private func disableLoginButton() {
+        loginButton.do {
+            $0.setLayer(borderColor: .gray4, borderWidth: 1)
+            $0.setTitleColor(.gray2, for: .normal)
+            $0.backgroundColor = .basicBlack
+            $0.isEnabled = false
+        }
+    }
+    
+    private func enableLoginButton() {
+        loginButton.do {
+            $0.setTitleColor(.basicWhite, for: .normal)
+            $0.backgroundColor = .brandRed
+            $0.layer.borderWidth = 0
+            $0.isEnabled = true
+        }
+    }
```
삼항연산자나 에러타입 확장성을 고려해서 enum으로 만드셔도 좋을듯 싶어여

JinUng41 replied:
말씀해 주신 부분은 삼항연산자를 통해 중복된 내용을 하나로 통합하고자 수정하였습니다.

seungchan2 commented on 34th-Sopt-Assignment/34th-Sopt-Assignment/Controllers/Login/LoginViewController.swift:
```diff
@@ -0,0 +1,406 @@
+        print(#function)
+        switch sender.tag {
+        case 0:
+            idTextField.text = nil
+            idTextField.insertText("")
```
idTextField.text = ""
랑 어떤게 다른건가요?!

JinUng41 replied:
단순히 textField.text = nil 또는 ""를 한다고 해서 이벤트가 호출되지 않습니다. 
`textFieldEditingChanged`가 다시 불리기 위해서는 텍스트필드의 `insertText("")`를 호출하여야 했기 때문에 다음과 같이 구현하였습니다.

seungchan2 commented on 34th-Sopt-Assignment/34th-Sopt-Assignment/Controllers/Login/LoginViewController.swift:
```diff
@@ -0,0 +1,406 @@
+        let viewController = MakeNicknameViewController()
+        viewController.delegate = self
+        viewController.modalPresentationStyle = .formSheet
+        if let sheet = viewController.sheetPresentationController {
+            sheet.do {
+                $0.detents = [.medium()]
+                $0.prefersGrabberVisible = true
+                $0.preferredCornerRadius = 24.0
+            }
+        }
```
SnapKit으로 sheet 구현해보셔도 좋을것 같네요! 적절한 애니메이션과 함께 ..

JinUng41 replied:
네 감사합니다.
추후 SnapKit으로도 구현해 보겠습니다!

seungchan2 commented on 34th-Sopt-Assignment/34th-Sopt-Assignment/Controllers/Login/LoginViewController.swift:
```diff
@@ -0,0 +1,406 @@
+            $0.textAlignment = .center
+        }
+        
+        idTextField.do {
+            $0.setText(
+                placeholder: "아이디",
+                textColor: .basicWhite,
+                backgroundColor: .gray4,
+                placeholderColor: .gray2,
+                font: .pretendard(weight: .six, size: 15)
+            )
+            $0.setAutoType()
+            $0.setLayer(borderColor: .basicWhite)
+            $0.addPadding(left: 20)
+            $0.keyboardType = .emailAddress
+            $0.rightView = idTextFieldRightView
+            $0.rightViewMode = .whileEditing
+            $0.addTarget(self, action: #selector(textFieldEditingChanged), for: .editingChanged)
+        }
```
UITextField를 클래스 분리하는건 어떨까요? 생성자 주입으로 ..?!

JinUng41 replied:
생성자에서 어떠한 것을 기준으로 받을 것인지 등에 대한 고민이 있습니다.
이를 판단하는 것은 제가 아직 많이 구현해보지 못한 경험에서 나오는 짬이 부족해서 아닐까 싶기도 하네요.. ㅎㅎㅎ
저는 뷰컨트롤러 최상단, UIComponent 선언 시에는 정말 컴포넌트가 어떤게 쓰이는지 확인하는 용도이며, 아래의 익스텐션에서 실제 내용을 덧붙이는 생각으로 구현을 했습니다.
이렇게 한 이유는 최상단에서 좀 더 비즈니스 로직에 집중할 수 있도록 가독성 측면을 고려해본건데요.
이러한 제 생각은 어떤지 같이 의논해 보고 싶어요!

seungchan2 replied:
> 지금 당장은 아이디, 비밀번호로 나눠도 좋을 것 같아요.
> 실제 티빙앱을 확인해보면 비슷한 텍스트필드가 여러 곳에서 쓰이는 것 같아요.
> 해당 UI 로직이 15줄 정도 되는 것 같은데 이게 쌓이다보면 보일러 플레이트 코드가 될 수도 있다는 생각이 듭니다!
> 
> ```
> private let idTextField = TvingTextField(type: .id)
> private let passwordTextField = TvingTextField(type: .password)
> ```
> 
> TvingTextFiled 내 케이스 별로 처리
> 
> 
> enum으로 해당 케이스들을 관리하고 UITextField에 해당 enum을 생성자로 사용한다면 중복되는 코드가 줄지 않을까 생각합니다!
> 
> + 비지니스 로직에 집중할 수 있도록 한다는 게 어떤 의미일까요?

JinUng41 replied:
비즈니스 로직에 집중한다는 것은 좀 더 화면에 데이터를 다루는 로직이라던지 화면 이동에 관한 로직이라던지 등 뷰 초기 설정 이외의 것들에 코드 상단에서부터 확인할 수 있으면 좋지 않을까 하는 의미였습니다.

말씀해주신 커스텀 UI 클래스 구성은 고민하여 적용해 보겠습니다.

seungchan2 commented on 34th-Sopt-Assignment/34th-Sopt-Assignment/Controllers/Login/LoginViewController.swift:
```diff
@@ -0,0 +1,406 @@
+        }
+        
+        idClearButton.snp.makeConstraints {
+            $0.width.height.equalTo(20)
```
$0.size로 한번에 잡을 수 있겠네요!

JinUng41 replied:
네 수정하였습니다.
전혀 모르고 있었던 부분이네요.

JinUng41 committed:
#1 - [Refactor] PR #2의 코드 리뷰에 따른 수정

- UIImage resource 사용으로 인한 Constants.Image 삭제
- 텍스트 필드 입력에 따른 버튼 활성화 코드 수정 (삼항 연산자 사용)
- catch let 구문을 활용한 오류 처리
- snapkit의 leading과 trailing을 horizontalEdges로 수정

JinUng41 committed:
#1 - [Config] Asset 및 폰트 파일 재추가

JinUng41 committed:
#1 - [Refactor] PR #2 코드 리뷰 2차 반영

- SetUI 및 AutoLayout의 extension을 private으로 수정
- snapkit의 width.height를 size로 수정
- 인자를 3개 이상 받는 메서드의 개행 처리

JinUng41 committed:
#1 - [Refactor] Constants를 enum으로 선언

JinUng41 committed:
#1 - [Feat] 텍스트필드 컴포넌트화

- TvingTextField 객체 구현

codeJiwon commented on 34th-Sopt-Assignment/34th-Sopt-Assignment/Controllers/Login/LoginViewController.swift:
```diff
@@ -0,0 +1,365 @@
+    
+    private let nicknameButton = UIButton()
+    
+    // MARK: - Property
```
진웅님 조언 듣고 mark 주석에 대해 찾아봤는데, 확실히 한눈에 어떤 내용인지 파악할 수 있어서 좋네요! 저도 이후 과제에 사용해보겠습니다

codeJiwon commented:
> ## What is the PR? 🔍
> * 2차 세미나 과제 (기본, 도전)
> * 로그인, 닉네임, 환영 화면 구현하기
> 
> ## Changes 📝
> ### 입력에 대한 에러 처리
> * 아이디(이메일), 비밀번호, 닉네임을 정규식을 통해 검증하는 부분을 에러를 `throw`할 수 있도록 구현
> * 에러 케이스를 정의하고 Alert를 띄울 수 있도록 하였습니다.
> 
> AppError 코드
> AlertShowable 코드
> ## Screenshot 📷
> 오류처리	정상동작
> ![](https://private-user-images.githubusercontent.com/91656206/323345454-f13c932d-0367-4ab6-8176-371ebd35d4bf.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM4ODU5MTYsIm5iZiI6MTcxMzg4NTYxNiwicGF0aCI6Ii85MTY1NjIwNi8zMjMzNDU0NTQtZjEzYzkzMmQtMDM2Ny00YWI2LTgxNzYtMzcxZWJkMzVkNGJmLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDIzVDE1MjAxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFmMzAxYzEyNzYyYzk0YWQ1Zjk1MzI0ZDI3NWU3ZmRjODM1MTkwZmI3YmZiNzU5NDBiZGQ1NTQ0ZWMzOGU5ZjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.YaAOiQNNkFGm33V2Ue_kl7Rpz6TGwZRn7RtN0cNh_LY)	![](https://private-user-images.githubusercontent.com/91656206/323345562-5eff4af2-76fe-4678-8f70-6a1e5d7d98c2.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM4ODU5MTYsIm5iZiI6MTcxMzg4NTYxNiwicGF0aCI6Ii85MTY1NjIwNi8zMjMzNDU1NjItNWVmZjRhZjItNzZmZS00Njc4LThmNzAtNmExZTVkN2Q5OGMyLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDIzVDE1MjAxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYzMGE4MmM2YTA0NWYyZGViMDk3ZTk4NDIzNjhlNTNiZTY4OWJiOTg4YTUzYmYxNjc5NjQxYjRmZThjZDNhNmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.xZO8QiN91FJ_T4D0xFqJxzuN0B-Cd_23dV9H7gkJgHU)
> ## To Reviewers 🙏
> * 커스텀 델리게이트 패턴을 이용한 데이터 전달 (닉네임)
> * 하드 코딩을 줄이기 위한 Constants 구현
> * MARK 주석 및 Then 적극 사용
> * 익스텐션을 이용한 코드 간략화
> 
> ## Related Issues 💭
> * Resolved: [[Feat] 2차 세미나 과제 구현 #1](https://github.com/NOW-SOPT-iOS-Part/KimJinUng-assignment/issues/1)

codeJiwon commented:
진웅님이 해주신 코드리뷰 보고 진웅님 코드도 참고하니까 제 코드에서 어떤 부분 개선시켜야할 지 배울 수 있어서 너무 좋았습니다! 정말 감사합니다

HELLOHIDI reviewed (commented):
정말 생각이 많이 담긴 코드라는게 느껴졌습니다. 코드리뷰의 선순환도 많이 느껴졌구요 너무너무 고생했습니다!
